src/estimagic/optimization/subsolvers/_steihaug_toint_quadratic.py

```diff
@@ -0,0 +1,209 @@
+            rr,
+            trustregion_radius,
+            norm_p,
+            n_iter,
+        )
+
+    for _ in range(max_iter):
+        alpha = rz / kappa
+        norm_dp1 = norm_d + alpha * (2 * dp + alpha * norm_p)
+
+        if trustregion_radius != 0 and norm_dp1 >= radius_sq:
+            reason = "Converged_CG_Constrained"
+            converged = True
+
+            if norm_p > 0:
+                x_candidate = _take_step_to_trustregion_boundary(
+                    x_candidate, p, dp, radius_sq, norm_d, norm_p
+                )
+
+            break
+
+        x_candidate = x_candidate + alpha * p
+        residual = residual - alpha * (model_hessian @ p)
+
+        norm_d = x_candidate.T @ x_candidate
+
+        rzm1 = rz
+        rz = residual.T @ residual
+
+        norm_r = np.linalg.norm(residual)
+
+        converged, diverged, reason = _check_convergence(
+            norm_r, norm_r0, abstol, ttol, divtol, converged, diverged, reason
+        )
+
+        if converged or diverged:
+            break
+
+        beta = rz / rzm1
+
+        if abs(beta) <= 0:
+            reason = "Diverged_Breakdown"
+            diverged = True
+            break
+
+        if n_iter >= max_iter:
+            reason = "Diverged_maxiter"
+            diverged = True
+            break
+
+        p = residual + beta * p
+
+        dp = x_candidate.T @ p
+        norm_p = p.T @ p
+
+        z = model_hessian @ p
+        kappa = p.T @ z
+        n_iter += 1
+
+        if kappa <= 0:
+            reason = "Converged_CG_NEG_CURVE"
+            converged = True
+
+            if trustregion_radius != 0 and norm_p > 0:
+                x_candidate = _take_step_to_trustregion_boundary(
+                    x_candidate, p, dp, radius_sq, norm_d, norm_p
+                )
+
+            break
+
+    return x_candidate
+
+
+def _update_candidate_vector_and_iteration_number(
+    x_candidate,
+    residual,
+    p,
+    z,
+    model_gradient,
+    model_hessian,
+    rr,
+    radius,
+    norm_p,
+    n_iter,
+):
+    """Update candidate, z vector, and iteration number."""
+    radius_sq = radius**2
+
+    if radius != 0 and norm_p > 0:
+        # Take step to boundary
+        step = np.sqrt(radius_sq / norm_p)
+        x_candidate = x_candidate + step * p
+
+    elif radius != 0:
+        if radius_sq >= rr:
+            alpha = 1.0
+        else:
+            alpha = np.sqrt(radius_sq / rr)
+
+        x_candidate = x_candidate + alpha * residual
+        z = model_gradient - 0.5 * (model_hessian @ x_candidate)
+
+        n_iter += 1
+
+    return x_candidate, z, n_iter
+
+
+def _take_step_to_trustregion_boundary(x_candidate, p, dp, radius_sq, norm_d, norm_p):
+    """Take step to trust-region boundary."""
+    step = (np.sqrt(dp * dp + norm_p * (radius_sq - norm_d)) - dp) / norm_p
+    x_candidate = x_candidate + step * p
+
+    return x_candidate
+
+
+def _check_convergence(
+    rnorm, rnorm0, abstol, ttol, divtol, converged, diverged, message
```

Comment:
Why does this take message as argument? It looks like the passed in message is not used and just redefined inside the function.

Reply:
Ok, I see now. In some cases the message is not changed and just returned. Same for the other things. Maybe add an else clause to make that explicit.

Reply:
Got rid of internal variable `reason` (which is called `message` in the function above).